source/classes.tex

```diff
@@ -606,8 +606,8 @@
 
 \pnum
 A class is regarded as complete
-where its definition is reachable and
-within its complete-class contexts;
+within its complete-class contexts and also
+wherever its definition is reachable;
 otherwise it is regarded as incomplete within its own class
 \grammarterm{member-specification}.
```

Comment:
It's a little strange for the "otherwise" case to not cover all cases not covered by the previous rule. This reordering makes it stranger by moving the complete-class context part away from the other member-secification pieces.

How about something like:

> A class is complete where its definition is reachable, and in complete-class contexts within its *member-specification*. Otherwise it is incomplete, including in its *member-specification*.

Reply:
Yeah, sounds better. Updated.

@tkoeppe?

Reply:
Very good, thank you!

Reply:
I think a constructive rule is better:
> A class is complete:
>> - at the point where its definition is reachable, or
>> - in complete-class contexts within its member-specification
>
> Otherwise it is incomplete, including in other parts of its member-specification.

Reply:
Why do you think that's better than the previous suggestion? The sentence seems short enough to not require a list, and also I don't think the "or" is right: a class is really complete in both cases, not just in "at least one of these cases". "Or" would be appropriate if we said, "In order to be complete, a class has to [either X] or [Y]", but as written, "and" seems more appropriate.

Reply:
Mike M. says that "complete-class context" is already applying to "within its member-specification" per the definition, so the extra normative mention of "within its member-specification" is superfluous.

However, this is now deviating from the existing words quite a bit, so I'm wondering whether we want a brief trip through CWG.

Reply:
I think the "within the member-specification" in otherwise case can be removed. We have completely defined what complete-class context is:
> A complete-class context of a class (template) is a 
>> - [...]
>>
>  **within the member-specification of the class or class template**.

This implies, e.g., that the function body of the member function **outside of the member-specification** is not a complete-class context. Hence, 
> the class is considered to be complete either:
>> - at the point where its definition is reachable, or
>> - in complete-class contexts within its member-specification
>>
> Otherwise, it is incomplete.  

I think this is clear.

Reply:
> a class is really complete in both cases, not just in "at least one of these cases". "Or" would be appropriate if we said, "In order to be complete, a class has to [either X] or [Y]", but as written, "and" seems more appropriate.

> For `and` in the updated wording, I still think "and" can result in ambiguity. How should we parse that wording?

Mind you, the current wording uses "and also", which is correct.

Reply:
Where does "and also" appear in the current wording? The current wording is just ambiguous since its "and".

Reply:
Oh, I was looking at the diff of this PR, which added "also".